{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e8ee1f84-e5c8-4bd8-af21-563767423ae5",
   "metadata": {},
   "source": [
    "# Primer BLASTing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "953a7c90-2d60-4eb8-a206-99cab4143422",
   "metadata": {},
   "source": [
    "## Installations \n",
    "\n",
    "- `sudo apt-get install -y ncbi-entrez-direct`\n",
    "- `sudo apt install -y ncbi-blast+`\n",
    "\n",
    "## Preparations\n",
    "\n",
    "- `mkdir CompBiol2023`\n",
    "- `cd CompBiol2023`\n",
    "\n",
    "### Download human chomosome 3\n",
    "\n",
    "- `efetch -db nuccore -id NC_000003 -format fasta > hs-chr3.fasta`\n",
    "\n",
    "### Download D3S1358 primer pair\n",
    "\n",
    "- goto: https://strbase-archive.nist.gov/str_D3S1358.htm\n",
    "- create FASTA file called *D3S1358.primers*\n",
    "\n",
    "```\n",
    ">f-primer\n",
    "ACTGCAGTCCAATCTGGGT\n",
    ">r-primer\n",
    "ATGAAATCAACAGAGGCTTG\n",
    "```\n",
    "\n",
    "## Look for primers in chromosome\n",
    "\n",
    "- `egrep --color 'ACTGCAGTCCAATCTGGGT' hs-chr3.fasta `\n",
    "- `egrep --color 'ATGAAATCAACAGAGGCTTG' hs-chr3.fasta `\n",
    "\n",
    "## Setup BLAST database\n",
    "\n",
    "- `makeblastdb -in hs-chr3.fasta -dbtype nucl -out hsc3`\n",
    "\n",
    "## BLASTing\n",
    "\n",
    "- `time blastn -query D3S1358.primers -db hsc3 -out d3s1358-vs-chr3.txt`\n",
    "- `less d3s1358-vs-chr3.txt`\n",
    "    - take a look into the file with `less`\n",
    "    - there are no hits do to short sequence BLASTing\n",
    "\n",
    "- `blastn -query D3S1358.primers -db hsc3 -out d3s1358-vs-chr3.txt -task blastn-short`\n",
    "    - adjust BLASTN algorithm\n",
    "    - take a look into the file\n",
    "\n",
    "- `blastn -query D3S1358.primers -db hsc3 -out d3s1358-vs-chr3.tab -task blastn-short -outfmt 6`\n",
    "    - adust BLAST output to tabular\n",
    "    - take a look into the file\n",
    "\n",
    "- `awk '$3==100 && $4>=19{print $1, $9, $10}' d3s1358-vs-chr3.tab`\n",
    "    - print only good matches\n",
    "\n",
    "## Retrieve amplicon sequence\n",
    "\n",
    "- `blastdbcmd -db hsc3 -entry all -range 45540713-45540843`\n",
    "- `blastdbcmd -db hsc3 -entry all -range 45540713-45540843 | sed '1d' | sed 's/TCTA/ & /g' | egrep --color '(ATGAAATCAACAGAGGCTTGC|ACCCAGATTGGACTGCAGT)'`\n",
    "    - highlight STR repeats\n",
    "\n",
    "## TASK\n",
    "Download chromosome 3 sequences for other individuals and orangutan (Pongo)\n",
    "\n",
    "- `efetch -db nuccore -id CM000553 -format fasta > hs-chr3-pongo.fasta`\n",
    "- `efetch -db nuccore -id AP023463 -format fasta > hs-chr3-japanese.fasta`\n",
    "- `efetch -db nuccore -id CM021570 -format fasta > hs-chr3-ashkenazi.fasta`\n",
    "- `efetch -db nuccore -id CH003498 -format fasta > hs-chr3-venter.fasta`\n",
    "- `efetch -db nuccore -id NC_000003 -format fasta > hs-chr3.fasta`\n",
    "\n",
    "Repeat the above proceedure and describe your observations\n",
    "\n",
    "---\n",
    "\n",
    "# Automating Primer Download\n",
    "\n",
    "## Installation\n",
    "\n",
    "- `sudo apt install pandoc`\n",
    "\n",
    "## Process the website\n",
    "\n",
    "- `pandoc -f html -t plain https://strbase-archive.nist.gov/PP16primers.htm`\n",
    "- `pandoc -f html -t plain https://strbase-archive.nist.gov/PP16primers.htm | egrep  '(Pair |[ATGC]{10,})' `\n",
    "- `pandoc -f html -t plain https://strbase-archive.nist.gov/PP16primers.htm | egrep  '(Pair |[ATGC]{10,})' | awk '{if (NR%3 == 1) {id=$1} else if (NR%3 == 2) {print \">f_\"id; print $0} else if (NR%3 == 0) {print \">r_\"id; print $0}}'`\n",
    "\n",
    "## Download all primer pairs into individual FASTA files\n",
    "\n",
    "- `pandoc -f html -t plain https://strbase-archive.nist.gov/PP16primers.htm | egrep  '(Pair |[ATGC]{10,})' | awk '{if (NR % 3 == 1) {id=$1} else if (NR % 3 == 2) {print \">f_\"id > id\".fasta\"; print $0 >> id\".fasta\"} else if (NR % 3 == 0) {print \">r_\"id >> id\".fasta\"; print $0 >> id\".fasta\"}}'`\n",
    "\n",
    "---\n",
    "\n",
    "# Automatation of the Process for Marker D21S11\n",
    "\n",
    "## Download chromosomes\n",
    "- `efetch -db nuccore -id CM000571 -format fasta > chr21-pongo.fasta`\n",
    "- `efetch -db nuccore -id AP023481 -format fasta > chr21-japanese.fasta`\n",
    "- `efetch -db nuccore -id CM021588 -format fasta > chr21-ashkenazi.fasta`\n",
    "- `efetch -db nuccore -id CH003516 -format fasta > chr21-venter.fasta`\n",
    "- `efetch -db nuccore -id NC_000021 -format fasta > chr21-ref.fasta`\n",
    "\n",
    "## Create BLAST databases\n",
    "\n",
    "- `for i in chr21*fasta; do base=$(basename $i .fasta); makeblastdb -in $i -dbtype nucl -out $base; done`\n",
    "\n",
    "## BLAST primer and create tabular output\n",
    "\n",
    "- `for i in chr21*fasta; do base=$(basename $i .fasta); blastn -query D21S11.fasta -db $base -out $base.tab -task blastn-short -outfmt 6; done`\n",
    "\n",
    "## Create function to extract primer pairs\n",
    "\n",
    "- `primerpair() { awk '{if($1~/f_/ && $3==100){forward[NR]=$9}else if($1~/r_/ && $3==100){reverse[NR]=$9}}END{for(x in forward){for(y in reverse){z=forward[x]-reverse[y]; if(z<0){z=z*-1}; if(z<400){gsub(/\\.tab/,\"\",FILENAME); print FILENAME, reverse[y],forward[x]}}}}' $1; }`\n",
    "\n",
    "## Use function to extract primer matches\n",
    "\n",
    "- `for i in chr21*tab; do primerpair $i; done`\n",
    "\n",
    "## Retrieve amplicon sequences\n",
    "\n",
    "- `for i in chr21*tab; do base=$(basename $i .tab); echo $i; primerpair $i | awk '{cmd=\"blastdbcmd -db \"$1\" -entry all -range \"$3\"-\"$2\" | sed '1d'\"; system(cmd)}'; done`\n",
    "\n",
    "- `for i in chr21*tab; do base=$(basename $i .tab); echo $i; primerpair $i | awk '{cmd=\"blastdbcmd -db \"$1\" -entry all -range \"$3\"-\"$2\" | sed '1d'\"; system(cmd)}' | grep TATC; done`\n",
    "    - extract repeat length and highlight\n",
    "\n",
    "- `for i in chr21*tab; do base=$(basename $i .tab); echo $i; primerpair $i | awk '{cmd=\"blastdbcmd -db \"$1\" -entry all -range \"$3\"-\"$2\" | sed '1d'\"; system(cmd)}' | grep -o TATC | wc -l; done`\n",
    "    - extract repeat length and count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b357b04a-cf35-4df4-b788-579a483b5489",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
